codalab/worker_manager/slurm_batch_worker_manager.py

```diff
@@ -0,0 +1,207 @@
+        )
+
+    def __init__(self, args):
+        super().__init__(args)
+        self.username = self.args.user if self.args.user else getpass.getuser()
+
+    def get_worker_jobs(self):
+        """
+        Return a list of workers.
+        The current Slurm version on NLP cluster is 17.11.13-2, which doesn't have rest api provided.
+        """
+        # Documentation can be found at https://slurm.schedmd.com/squeue.html.
+        # Get all the Slurm workers that are owned by the current user.
+        # Returning result will be in the following format:
+        # JOBID:STATE (header won't be included with "-h" option)
+        # 1478828:PENDING
+        # 1478830:PENDING
+        jobs = self.run_command(['squeue', '-u', self.username, '-o', '%i:%T', '-h'])
+        jobs = jobs.strip().split()
+        logger.info('Workers: {}'.format(' '.join(job for job in jobs) or '(none)'))
+
+        # Get all the RUNNING jobs that are owned by the current user.
+        # Returning result will be in the following format:
+        # JOBID (header won't be included with "-h" option)
+        # 1478828
+        # 1478830
+        running_jobs = self.run_command(
+            ['squeue', '-u', self.username, '-t', 'RUNNING', '-o', '%i', '-h']
+        )
+        running_jobs = running_jobs.strip().split()
+
+        return [WorkerJob(job) for job in running_jobs]
+
+    def start_worker_job(self):
+        """
+        Start a CodaLab Slurm worker that submits batch job to Slurm
+        """
+        worker_id = uuid.uuid4().hex
+        # Set user's local home directory as worker directory
+        work_dir = Path(str(Path.home()), "slurm-batch-scratch/{}".format(worker_id))
+        # Set up worker directory
+        work_dir.mkdir(parents=True, exist_ok=True)
+
+        # This needs to be a unique directory since Batch jobs may share a host
+        worker_network_prefix = 'cl_worker_{}_network'.format(worker_id)
+        command = [
+            'cl-worker',
+            '--server',
+            self.args.server,
+            '--verbose',
+            '--exit-when-idle',
+            '--idle-seconds',
+            str(self.args.worker_idle_seconds),
+            '--work-dir',
+            str(work_dir),
+            '--id',
+            worker_id,
+            '--network-prefix',
+            worker_network_prefix,
+            # always set in Slurm worker manager to ensure safe shutdown
+            '--pass-down-termination',
+            'password-file',
+            self.args.password_file,
+        ]
+        if self.args.worker_tag:
+            command.extend(['--tag', self.args.worker_tag])
+
+        slurm_args = self.create_slurm_args(self.args)
+        job_definition = self.create_job_definition(slurm_args=slurm_args, command=command)
+
+        # Not submit job to Slurm if dry run is specified
+        if self.args.dry_run:
+            return
+
+        batch_script = str(work_dir.joinpath(slurm_args['job-name'] + '.slurm'))
```

Comment:
Does this ever get cleaned up?

Reply:
Hmm, not currently. I thought it would be nice to keep those configuration locally for future reference. But we can definitely clean it up if that's prefered.

Reply:
+1 to cleaning up

Reply:
I think we should be saving the `batch_script` to a different directory than the `work-dir`, since the `work-dir` might not exist on the host running the `slurmworkermanager`. E.g., if we run the `slurmworkermanager` on sc, but we want to write to each worker's local disk (a path that doesn't exist on sc)